src/main/java/me/contaria/seedqueue/mixin/client/WorldGenerationProgressTrackerMixin.java

```diff
@@ -0,0 +1,85 @@
+package me.contaria.seedqueue.mixin.client;
+
+import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import me.contaria.seedqueue.SeedQueue;
+import me.contaria.seedqueue.interfaces.SQWorldGenerationProgressTracker;
+import net.minecraft.client.gui.WorldGenerationProgressTracker;
+import net.minecraft.server.WorldGenerationProgressLogger;
+import net.minecraft.util.Util;
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.chunk.ChunkStatus;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+@Mixin(WorldGenerationProgressTracker.class)
+public abstract class WorldGenerationProgressTrackerMixin implements SQWorldGenerationProgressTracker {
+    @Shadow @Final private Long2ObjectOpenHashMap<ChunkStatus> chunkStatuses;
+    @Shadow @Final private WorldGenerationProgressLogger progressLogger;
+    @Shadow private ChunkPos spawnPos;
+    @Shadow @Final private int radius;
+
+    @Unique private long creationTime = -1;
+
+    @Unique private boolean frozenStatesGathered = false;
+    @Unique private Long2ObjectOpenHashMap<ChunkStatus> frozenChunkStatuses;
+    @Unique private int frozenProgressPercentage;
+    @Unique private ChunkPos frozenSpawnPos;
+
+    @Unique private boolean unfrozen = false;
+
+    @Inject(method = "start()V", at = @At("TAIL"))
+    private void onStart(CallbackInfo ci) {
+        this.creationTime = Util.getMeasuringTimeMs();
+    }
```

Comment:
sure, I'll try to find that

Reply:
no start logging message, and if we go based on the first "preparing spawn area: x%" message, it looks like this even with 0ms delay
![image](https://github.com/user-attachments/assets/6b8f9c3a-ff2b-4779-9649-6ca38a30cd08)

Reply:
Ah it seems `start` is basically right after the "Preparing start region" log message, so that should be fine then
![grafik](https://github.com/user-attachments/assets/a98b56c5-cf13-42c5-80a8-30fbfc81410c)

Reply:
Actually, `worldGenerationProgressListener` is a `QueueingWorldGenerationProgressListener`, which queues the start to be done on the client thread, so there is some delay

Reply:
Probably still fine tho

Reply:
The [logic from a village resetting julti fork](<https://github.com/Avzes1/Julti/commit/93753ee1e1fe02c91ccd84aba553fe6c1d44aa4c#diff-6fc6dbe51230337a5c8faa14fd4aeb3ca76946df0bab07870ca3b642579328a2R215>) is basically `currentState == "generating" and lastStartOfGeneratingState >= freezeDelayMs`

Reply:
> Injecting right after the LOGGER.info line in prepareStartRegion with 0ms freezing gives this ![image](https://private-user-images.githubusercontent.com/59705125/361511570-2be68f8a-0e70-47e3-ad6d-a7b6da227e82.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2OTQ3MzEsIm5iZiI6MTcyNDY5NDQzMSwicGF0aCI6Ii81OTcwNTEyNS8zNjE1MTE1NzAtMmJlNjhmOGEtMGU3MC00N2UzLWFkNmQtYTdiNmRhMjI3ZTgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDE3NDcxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIzZmNmMWMyNWIzMTdhMjA1NjMzMjJiNzNhY2QwOWNlNTUxN2M5NjM5ZDU4YjhlMTg4MjE1Nzg0NTlmODY3N2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0mYgs16ZQ-SKGNc5T115qsaHn9ypZDfzcw0xWgdiGVM) (seed is set to a village seed)

Actually this makes sense, spawnpoint is found earlier than the `prepareStartRegion` method is called and I think `WorödGenerationProgressTracker#start` is set to 0 0 before it's set to the actual spawn. StateOutput also hooks into the tracker so starting the timer on `start` is definitely correct.

Reply:
state output only hooks into the progress tracker for the percentages, the initial `generating,0` happens when a LevelLoadingScreen is constructed

Reply:
The only place this is constructed seems to be in startIntegratedServer, and it has access to the world gen progress tracker directly, even without having to access it through the QueueingWorldGenerationProgressListener

Reply:
actually the tracker is given directly in the constructor for level loading screen so I can inject to the tail of that

Reply:
Oh i see, so thats what you mean by it not working correctly outside of wall.
You could inject into `<init>` or add to `MinecraftClientMixin#saveWorldGenerationProgressTracker` and schedule a client runnable (`SeedQueue#scheduleTaskOnClientThread`) to set start time.
However because the wall screen is obviously more laggy and more prone to lag spikes than just loading a single preview, there could occasinally be a much bigger delay between the schedule and execution than youd get in traditional wall.
So while implementing the server-client delay is more accurate, the delay is potentially much bigger than usual.

Reply:
ngl I have no idea what you're talking about

`generating,0` happens ok level loading screen construction so that's where I've marked the start time to base the freezing off of

Reply:
Okay disregard what I said about scheduling it to be run on the client.

The problem at hand is that the current approach doesnt work when not on the wall screen or with too low background previews like you said.
In `startIntegratedServer`, the client waits for the tracker to be set by the server, then creates the `LevelLoadingScreen` with that. The tracker is immediately set after its been constructed so I think its fair to just start the cooldown at the TAIL of `WorldGenerationProgressTracker<init>`